File: Leetcode/Vertical Order Traversal of a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def verticalTraversal(self, root: TreeNode) -> List[List[int]]:
        dic = defaultdict(list)
        self.vertOrder(root, dic, 0, 0)
        res = []
        for i in sorted(dic.keys()):
            tmp = []
            for j in sorted(dic[i]):
                tmp.append(j[1])
            res.append(tmp)
        return res

    def vertOrder(self, root, dic, vl, hl):
        if not root:
            return
        dic[hl].append((vl, root.val))
        self.vertOrder(root.left, dic, vl+1, hl-1)
        self.vertOrder(root.right, dic, vl+1, hl+1)

File: Leetcode/Vertical Order Traversal of a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
    map<int, set<pair<int, int>>> mp;
    void verticalOrder(TreeNode *root, int vl, int hl)
    {
        if (!root)
            return;
        mp[hl].insert({vl, root->val});
        verticalOrder(root->left, vl + 1, hl - 1);
        verticalOrder(root->right, vl + 1, hl + 1);
    }

public:
    vector<vector<int>> verticalTraversal(TreeNode *root)
    {

        verticalOrder(root, 0, 0);
        vector<vector<int>> ans;
        for (auto m : mp)
        {
            vector<int> v;
            for (auto e : m.second)
                v.push_back(e.second);
            ans.push_back(v);
        }
        return ans;
    }
};